bigtable/client/polling_policy.h

```diff
@@ -0,0 +1,113 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+
+#include <grpc++/grpc++.h>
+
+#include "bigtable/client/rpc_backoff_policy.h"
+#include "bigtable/client/rpc_retry_policy.h"
+#include "bigtable/client/version.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Define the interface for providing asynchronous repetitive call rules
+ *
+ */
+class PollingPolicy {
+ public:
+  virtual ~PollingPolicy() = default;
+
+  /**
+   * Return a new copy of this object.
+   *
+   * Typically implemented as
+   * @code
+   *   return std::unique_ptr<PollingPolicy>(new Foo(*this));
+   * @endcode
+   */
+  virtual std::unique_ptr<PollingPolicy> clone() = 0;
+
+  /**
+   * Return true if `status` represents a permanent error that cannot be
+   * retried.
+   */
+  virtual bool IsPermanentError(grpc::StatusCode const& status_code) = 0;
+
+  /**
+   * Return true if we cannot try again.
+   */
+  virtual bool Exhausted() = 0;
+
+  /**
+   * Return for how long we should wait before trying again.
+   */
+  virtual std::chrono::milliseconds WaitPeriod() = 0;
+};
+
+// Define the defaults using a pre-processor macro, this allows the application
+// developers to change the defaults for their application by compiling with
+// different values.
+#ifndef BIGTABLE_CLIENT_DEFAULT_MAXIMUM_RETRY_PERIOD
+#define BIGTABLE_CLIENT_DEFAULT_MAXIMUM_RETRY_PERIOD std::chrono::minutes(6)
+#endif  // BIGTABLE_CLIENT_DEFAULT_MAXIMUM_RETRY_PERIOD
+
+const auto maximum_retry_period = BIGTABLE_CLIENT_DEFAULT_MAXIMUM_RETRY_PERIOD;
+
+#ifndef BIGTABLE_CLIENT_DEFAULT_INITIAL_DELAY
+#define BIGTABLE_CLIENT_DEFAULT_INITIAL_DELAY std::chrono::milliseconds(10)
+#endif  // BIGTABLE_CLIENT_DEFAULT_INITIAL_DELAY
+
+#ifndef BIGTABLE_CLIENT_DEFAULT_MAXIMUM_DELAY
+#define BIGTABLE_CLIENT_DEFAULT_MAXIMUM_DELAY std::chrono::minutes(5)
+#endif  // BIGTABLE_CLIENT_DEFAULT_MAXIMUM_DELAY
+
+const auto default_initial_delay = BIGTABLE_CLIENT_DEFAULT_INITIAL_DELAY;
+const auto default_maximum_delay = BIGTABLE_CLIENT_DEFAULT_MAXIMUM_DELAY;
+
+template <typename Retry = LimitedTimeRetryPolicy,
+          typename Backoff = ExponentialBackoffPolicy>
+class GenericPollingPolicy : public PollingPolicy,
+                             private Retry,
+                             private Backoff {
+ public:
+  GenericPollingPolicy()
+      : Retry(maximum_retry_period),
+        Backoff(default_initial_delay, default_maximum_delay) {}
```

Comment:
Ahh, you need the defaults in the header because the classes are templates.  To suggestions:

- Create functions (declared here, implemented in the `.cc` file), that return the default values.
- Or (better I think) add a default constructor for ExponentialBackoffPolicy and LimitedTimeRetryPolicy that initialize (in the `.cc` file) with the right values, from here you just call the default constructors.